openoffload/go/server/sessionoffload_thread.go

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022 Intel Corporation, or its subsidiaries.
+//
+// This file contains the logic for the session_update thread. This is meant to
+// simulate the random updating of sessions in the session table and would not
+// be used in a setup on actual hardware.
+//
+
+package main
+
+import (
+	"log"
+	"math/rand"
+	"time"
+
+	fw "github.com/opiproject/sessionOffload/sessionoffload/v2/gen/go"
+)
+
+// Update packet counters
+func session_update_packet_counters(v *session) {
+	// Increment packet counters
+	v.in_packets  += uint64(rand.Intn(1000))
+	v.out_packets += uint64(rand.Intn(1000))
+	v.in_bytes    += uint64(rand.Intn(100000))
+	v.out_bytes   += uint64(rand.Intn(100000))
+}
+
+func session_timeout(v *session) {
+	v.session_state      = fw.SessionState__CLOSED
+	v.session_close_code = fw.SessionCloseCode__TIMEOUT
+	v.end_time           = time.Now()
+}
+
+func session_fin(v *session) {
+	v.session_state      = fw.SessionState__CLOSED
+	v.session_close_code = fw.SessionCloseCode__FINACK
+	v.end_time           = time.Now()
+}
+
+// Function which runs in the background updating the session table entries
+func session_update() {
+	states := [18]string  {
+		"noop", "noop", "noop", "noop", "noop", "noop", "noop", "noop", "noop", "noop",
+		"stats", "stats", "stats", "stats", "stats",
+		"timeout", "clientfin", "serverfin",
+	}
+
+	for {
+		log.Printf("----- session_update running -----")
+		time.Sleep(time.Duration(*update) * time.Second)
+
+
+		for k, v := range sessions {
+			session_lock.RLock()
```

Comment:
Is it safe to iterate over `sessions` without holding the lock outside the loop?

Reply:
Another good catch!